File: DS/Java/AVLTree.java
import java.util.*;

public class AVL {
  Node root;

  private class Node {
    Node left;
    Node right;
    int data;

    Node(int d) {
      this.data = d;
      this.left = null;
      this.right = null;
    }
  }

  private int getHeight(Node head) {
    if (head == null)
      return 0;
    return 1 + Math.max(getHeight(head.left), getHeight(head.right));
  }

  private int getBalance(Node head) {
    return this.getHeight(head.left) - this.getHeight(head.right);
  }

  private Node rotateRight(Node head) {
    Node h = head.left;
    Node t = h.right;
    head.left = t;
    h.right = head;
    return h;
  }

  private Node rotateLeft(Node head) {
    Node h = head.right;
    Node t = h.left;
    head.right = t;
    h.left = head;
    return h;
  }

  private Node insertR(int k, Node head) {
    if (head == null)
      return new Node(k);
    if (head.data > k) {
      head.left = insertR(k, head.left);
    } else {
      head.right = insertR(k, head.right);
    }
    int b = this.getBalance(head);
    if (b > 1) {
      if (this.getBalance(head.left) < 0)
        head.left = this.rotateLeft(head.left);
      head = this.rotateRight(head);
    } else if (b < -1) {
      if (this.getBalance(head.right) > 0)
        head.right = this.rotateRight(head.right);
      head = this.rotateLeft(head);
    }
    return head;
  }

  public void insert(int k) {
    this.root = this.insertR(k, this.root);
  }

  public void levelOrder() {
    Queue<Node> q = new LinkedList<>();
    q.add(this.root);
    q.add(null);
    Node head;
    while (!q.isEmpty()) {
      Node temp = q.remove();
      if (temp == null && !q.isEmpty()) {
        q.add(null);
        System.out.println();
      } else if (temp != null) {
        if (temp.left != null)
          q.add(temp.left);
        if (temp.right != null)
          q.add(temp.right);
        System.out.print(temp.data + " ");
      }
    }
  }

  private void inorderR(Node head) {
    if (head == null)
      return;
    inorderR(head.left);
    System.out.print(head.data + " ");
    inorderR(head.right);
  }

  public void inorder() {
    this.inorderR(this.root);
    System.out.println();
  }

  public static void main(String[] args) {
    AVL t = new AVL();
    t.insert(5);
    t.insert(4);
    t.insert(3);
    t.insert(2);
    t.insert(1);
    t.inorder();
    t.levelOrder();
  }
}

File: APC/AVLtree.c
#include <stdio.h>
#include <stdlib.h>

#define MAX(X, Y) (X > Y) ? X : Y

typedef struct node
{
    int data;
    int height;
    struct node *left;
    struct node *right;
} node;

node *createNode(int data)
{
    node *temp = (node *)malloc(sizeof(node));
    temp->data = data;
    temp->left = NULL;
    temp->right = NULL;
    return temp;
}

node *leftRotate(node *x)
{
    node *t = x->right;
    node *u = t->left;
    t->left = x;
    x->right = u;
}

node *rightRotate(node *x)
{
    node *t = x->left;
    node *u = t->right;
    t->right = x;
    x->left = u;
}

int height(node *root)
{
    if (root == NULL)
        return 0;
    return 1 + MAX(height(root->left), height(root->right));
}

int getBalance(node *root)
{
    return height(root->left) - height(root->right);
}

node *insert(node *root, int key)
{
    if (root == NULL)
        return createNode(key);
    if (key < root->data)
        root->left = insert(root->left, key);
    else
        root->right = insert(root->right, key);
    root->height = MAX(height(root->left), height(root->right));
    int b = getBalance(root);
    if (b > 1)
    {
        if (getBalance(root->left) >= 0)
            return rightRotate(root);
        else
            root->left = leftRotate(root->left);
        return rightRotate(root);
    }
    else if (b < -1)
    {
        if (getBalance(root->right) >= 0)
            return leftRotate(root);
        else
            root->left = rightRotate(root->left);
        return rightRotate(root);
    }
}

void recursiveInorder(node *root)
{
    if (root == NULL)
        return;
    recursiveInorder(root->left);
    printf("%d ", root->data);
    recursiveInorder(root->right);
}

int main()
{
    node *root = NULL;
    root = insert(root, 1);
    root = insert(root, 2);
    root = insert(root, 3);
    root = insert(root, 4);
    root = insert(root, 5);
    recursiveInorder(root);
    // root->left = createNode(3);
}